src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -22,6 +22,87 @@ private enum EscapeLevel : uint
             ReturnOnly = ReturnOnlyScope,
         }
 
+        /// <summary>
+        /// Encapsulates a property or indexer and the accessor(s) on it that were used. The particular
+        /// indexer used is important as it impacts ref safety analysis
+        /// </summary>
+        private readonly struct MethodInfo
+        {
+            internal Symbol Symbol { get; }
+            internal MethodSymbol? Method { get; }
+            internal MethodSymbol? SetMethod { get; }
+            internal bool IsCompoundUsage => SetMethod is not null;
+            internal bool UseUpdatedEscapeRules => Method?.UseUpdatedEscapeRules == true;
+            internal bool ReturnsRefToRefStruct =>
+                Method is { RefKind: not RefKind.None, ReturnType: { } returnType } &&
+               returnType.IsRefLikeOrAllowsRefLikeType();
+
+            private MethodInfo(Symbol symbol, MethodSymbol? method, MethodSymbol? setMethod)
+            {
+                Symbol = symbol;
+                Method = method;
+                SetMethod = setMethod;
+            }
+
+            internal static MethodInfo Create(MethodSymbol method)
+            {
+                return new MethodInfo(method, method, null);
+            }
+
+            internal static MethodInfo Create(PropertySymbol property)
+            {
+                return new MethodInfo(
+                    property,
+                    property.GetMethod ?? property.SetMethod,
+                    null);
+            }
+
+            internal static MethodInfo Create(Symbol symbol, BoundExpression? expr)
+            {
+                if (symbol is MethodSymbol method)
+                {
+                    return new MethodInfo(symbol, method, setMethod: null);
+                }
+
+                if (symbol is PropertySymbol property)
+                {
+                    if (expr is BoundIndexerAccess indexer)
+                    {
+                        return Create(indexer);
+                    }
+
+                    // For properties the compiler prefers the `get` method over the `set` irrespective
+                    // of which one is actually used in the code. This simplification does lead to 
+                    // one case where the compiler incorrectly flags code as a ref safety error 
+                    // when it's not (when a readonly set is involved). That is tracked by the following
+                    // issue.
+                    //
+                    // If that issue raises to a high enough priority then likely it should take a similar
+                    // approach as BoundIndexerAccess.AccessKind
+                    //
+                    // https://github.com/dotnet/roslyn/issues/73872
+                    return new MethodInfo(
+                        symbol,
+                        property.GetOwnOrInheritedGetMethod() ?? property.GetOwnOrInheritedSetMethod(),
```

Comment:
With this change, do we have correct behavior in a case similar to https://github.com/dotnet/roslyn/issues/73872#issue-2338649632 but with different placement of 'readonly' modifiers:

```cs
ref struct S 
{
    internal Span<int> Span 
    {
        readonly get => default;
        set { }
    }

    readonly void M(Span<int> span) { }

    static void Test()
    {
        // safe to escape from method
        S local = new S();

        // stack alloc
        Span<int> span = [1, 2];

        // Error!
        local.Span = span;

        // Okay
        local.M(span);
    }
}
```

Reply:
Yes this works today because assignment validation is done on a simple escape basis. Basically is the escape scope of the LHS <= escape scope of the RHS. That means it doesn't actually look at the accessors here at all for whether or not they can mutate the receiver. Added a number of test cases to demonstrate this is done correctly. 

That does mean though we could probably solve the linked bug without having to go through `AccessKind` fully. Could probably just dig into the assignment validation a bit. Again though, seems low pri due to `readonly set` being a bit of an anti-pattern here.